core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java

```diff
@@ -22,68 +18,87 @@
  * frontiers.
  * @author manthan
  */
-public class BidirectionalSearch<A, S> implements SearchForActionsBidirectionallyFunction<A, S> {
+public class BidirectionalSearch<A, S> implements SearchForActionsBidirectionallyFunction<A, S>, BidirectionalActions<A> {
 
+    private List<A> fromInitialStatePartList = new ArrayList<>();
+    private List<A> fromGoalStatePartList = new ArrayList<>();
+    private List<A> fromInitialStateToGoalStateList = new ArrayList<>();
+    private List<A> fromGoalStateToInitialStateList = new ArrayList<>();
+    private BidirectionalActions<A> bidirectionalActions;
+    private Node<A, S> previousMeetingOfTwoFrontiers;
+    private Node<A, S> meetingOfTwoFrontiers;
+    private Node<A, S> nextNodeToBeEvaluated;
+    private boolean fromFront;
     @Override
-    public List<A> apply(Problem<A, S> originalProblem, Problem<A, S> reverseProblem) {
+
+    public BidirectionalActions<A> apply(Problem<A, S> originalProblem, Problem<A, S> reverseProblem) {
         Node<A, S> node = newRootNode(originalProblem.initialState(), 0);
         if (originalProblem.isGoalState(node.state())) {
-            return solution(node);
+            this.previousMeetingOfTwoFrontiers = node;
+            this.nextNodeToBeEvaluated = null;
+            return bidirectionalActions;
         }
         Node<A, S> revNode = newRootNode(reverseProblem.initialState(), 0);
         Queue<Node<A, S>> front = newFIFOQueue(node);
         Queue<Node<A, S>> back = newFIFOQueue(revNode);
-        Set<S> exploredFront = newExploredSet();
-        Set<S> exploredBack = newExploredSet();
+        HashMap<S, Node<A, S>> exploredFront = newExploredMap();
+        HashMap<S, Node<A, S>> exploredBack = newExploredMap();
```

Comment:
These should be 'Map<..>' and not 'HashMap<...>' - i.e. avoid restricting to a specific implementation.